### PR TITLE
[Agent] Add guardrail system for prompt injection detection and content validation

### LINKED ISSUE #1524: [Agent][Platform] Prompt injection detection/protection
Is „prompt injection detection/protection“ a topic that should/could be addressed as part of this project?

I had a quick chat with Claude today and maybe it is a good idea to port the „LLM Guard“ from Python or at least use it as inspiration.

I am working on a project that will provide an AI assistant based on Symfony AI. So, I will probably implement a solution anyway in the upcoming 3 months or so. If it makes sense to you, I could work on a pull request when I’m on it.

### PR DESCRIPTION
**Pull request by daifma**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Docs?         | yes
| Issues        | Fix #1524
| License       | MIT

## Description

This PR adds a modular guardrail system to the Agent component, enabling input and output validation through the existing processor pipeline. The design follows the tripwire pattern inspired by OpenAI's Agents SDK guardrails, adapted to Symfony AI's `InputProcessor`/`OutputProcessor` architecture.

### Architecture

The guardrail system introduces two interfaces:

- **`InputGuardrailInterface`** — validates input before it reaches the LLM
- **`OutputGuardrailInterface`** — validates output after the LLM responds

These are plugged into the agent pipeline via:

- **`GuardrailInputProcessor`** (implements `InputProcessorInterface`)
- **`GuardrailOutputProcessor`** (implements `OutputProcessorInterface`)

When a guardrail is triggered, a `GuardrailException` is thrown containing the `GuardrailResult` with details (scanner name, reason, confidence score).

### Built-in Scanners

Three scanners are included out of the box:

1. **`PromptInjectionScanner`** — Detects common prompt injection patterns (instruction overrides, role hijacking, system prompt extraction, jailbreak attempts). Ships with 11 default regex patterns and supports custom patterns.

2. **`InvisibleCharacterScanner`** — Detects hidden Unicode characters (zero-width spaces, bidirectional overrides, soft hyphens, etc.) that can be used to smuggle hidden instructions. Configurable threshold.

3. **`RegexScanner`** — User-defined regex pattern matching that works on both input and output (implements both interfaces). Useful for content filtering, PII detection, or custom validation rules.

### Usage Example

```php
use Symfony\AI\Agent\Agent;
use Symfony\AI\Agent\Guardrail\GuardrailInputProcessor;
use Symfony\AI\Agent\Guardrail\GuardrailOutputProcessor;
use Symfony\AI\Agent\Guardrail\Scanner\PromptInjectionScanner;
use Symfony\AI\Agent\Guardrail\Scanner\InvisibleCharacterScanner;
use Symfony\AI\Agent\Guardrail\Scanner\RegexScanner;

// Create scanners
$injectionScanner = new PromptInjectionScanner();
$invisibleScanner = new InvisibleCharacterScanner();
$piiScanner = new RegexScanner(
    patterns: ['/\b\d{3}-\d{2}-\d{4}\b/'],  // SSN pattern
    reason: 'PII detected in content',
);

// Wire into the agent pipeline
$inputProcessor = new GuardrailInputProcessor([$injectionScanner, $invisibleScanner, $piiScanner]);
$outputProcessor = new GuardrailOutputProcessor([$piiScanner]);

$agent = new Agent(
    platform: $platform,
    model: $model,
    inputProcessors: [$inputProcessor],
    outputProcessors: [$outputProcessor],
);
```

### Test Coverage

55 tests with 104 assertions covering:
- All three scanners with positive and negative cases
- Both processor classes (pass-through and tripwire behavior)
- Edge cases: image-only messages, custom patterns, invalid regex validation
- Data provider patterns for comprehensive injection pattern coverage

### CODE REVIEW

File: src/agent/src/Exception/GuardrailException.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * @author Abderrahman Daif <daif.abderrahman@gmail.com>
+ */
+final class GuardrailException extends RuntimeException
+{
+    public function __construct(
+        private readonly GuardrailResult $guardrailResult,
+        ?\Throwable $previous = null,
+    ) {
+        parent::__construct(
```
Comment (OskarStark) [inline src/agent/src/Exception/GuardrailException.php]: All on one line please and use named argument for $previous to avoid the 0

File: src/agent/CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,13 @@ CHANGELOG
 0.4
 ---
 
+ * Add guardrail system for prompt injection detection and content validation
+ * Add `InputGuardrailInterface` and `OutputGuardrailInterface` contracts
+ * Add `GuardrailInputProcessor` and `GuardrailOutputProcessor` to plug guardrails into the agent pipeline
+ * Add `PromptInjectionScanner` to detect common prompt injection patterns
+ * Add `InvisibleCharacterScanner` to detect hidden Unicode characters
+ * Add `RegexScanner` for user-defined pattern matching on input and output
+ * Add `GuardrailException` thrown when a guardrail tripwire is triggered
```
Comment (OskarStark) [inline src/agent/CHANGELOG.md]:
```suggestion
```

File: src/agent/src/Guardrail/GuardrailInputProcessor.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * @throws GuardrailException when a guardrail tripwire is triggered
+     */
+    public function processInput(Input $input): void
+    {
+        foreach ($this->guardrails as $guardrail) {
+            $result = $guardrail->validateInput($input);
+
+            if ($result->isTriggered()) {
```
Comment (OskarStark) [inline src/agent/src/Guardrail/GuardrailInputProcessor.php:44]: Why do we need this? Would it be sufficient to throw in the validate method? Just asking

File: src/agent/src/Guardrail/GuardrailOutputProcessor.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * @throws GuardrailException when a guardrail tripwire is triggered
+     */
+    public function processOutput(Output $output): void
+    {
+        foreach ($this->guardrails as $guardrail) {
+            $result = $guardrail->validateOutput($output);
+
+            if ($result->isTriggered()) {
```
Comment (OskarStark) [inline src/agent/src/Guardrail/GuardrailOutputProcessor.php:44]: Same for this here

File: src/agent/src/Guardrail/GuardrailResult.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+     * @param bool        $triggered whether the guardrail tripwire was triggered
+     * @param string|null $reason    the reason the guardrail was triggered
+     * @param float       $score     a confidence score between 0.0 and 1.0
+     * @param string      $scanner   the name of the scanner that produced this result
+     */
+    public function __construct(
+        private readonly bool $triggered,
+        private readonly ?string $reason = null,
+        private readonly float $score = 0.0,
+        private readonly string $scanner = '',
```
Comment (OskarStark) [inline src/agent/src/Guardrail/GuardrailResult.php]: Instead of an empty string I would go with ?string and default null

File: src/agent/src/Guardrail/Scanner/RegexScanner.php [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+/**
+ * Scans input and output text against user-defined regex patterns.
+ *
+ * This scanner can be used to detect custom patterns such as banned topics,
+ * restricted keywords, or domain-specific injection techniques.
+ *
+ * @author Abderrahman Daif <daif.abderrahman@gmail.com>
+ */
+final class RegexScanner implements InputGuardrailInterface, OutputGuardrailInterface
```
Comment (OskarStark) [inline src/agent/src/Guardrail/Scanner/RegexScanner.php:30]: While reading this class name, k think instead of "guardrail", "scanner" is more appropriate and self explanatory.

WDYT @chr-hertel ?

**Comment by daifma** [comment]
Regarding the architectural question about `GuardrailInputProcessor` / `GuardrailOutputProcessor`:

The processor exists mainly as a **grouping mechanism** — it lets you register multiple scanners as a single `InputProcessorInterface` on the agent, rather than registering each scanner individually as a separate processor.

That said, I agree it could be simplified. If you prefer, the scanners could throw `GuardrailException` directly in their validate methods (with `void` return type), and the processor would just be a thin loop with no result checking. This would remove `GuardrailResult` as a return type.

Happy to refactor either way — let me know your preference.

File: src/agent/src/Guardrail/Scanner/RegexScanner.php [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    /**
+     * @param list<string> $patterns regex patterns to match against
+     * @param string       $reason   the reason message when a pattern matches
+     */
+    public function __construct(
+        array $patterns,
+        private readonly string $reason = 'Content matched a restricted pattern.',
+    ) {
+        foreach ($patterns as $pattern) {
+            if (false === @preg_match($pattern, '')) {
```
Comment (Guikingone) [inline src/agent/src/Guardrail/Scanner/RegexScanner.php:48]: I would use `UnicodeString` for any string-related operations (especially on this type of condition where we're silencing the error), if possible, to apply on others scanners

File: src/agent/src/Guardrail/InputGuardrailInterface.php [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ * Validates agent input before it is sent to the LLM.
+ *
+ * Implementations scan user messages for potential threats such as
+ * prompt injections, invisible characters, or restricted content.
+ *
+ * @author Abderrahman Daif <daif.abderrahman@gmail.com>
+ */
+interface InputGuardrailInterface
+{
+    public function validateInput(Input $input): GuardrailResult;
```
Comment (Guikingone) [inline src/agent/src/Guardrail/InputGuardrailInterface.php:26]: What about `validate`? For both interfaces, we already know the type received as an argument, IMHO, `validate` is expressive enough, no need to specific the suffix 🤔

Comment (OskarStark) [reply src/agent/src/Guardrail/InputGuardrailInterface.php:26]: Then you cannot implement both interfaces in the same class

File: src/agent/src/Guardrail/GuardrailInputProcessor.php [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * @param iterable<InputGuardrailInterface> $guardrails
+     */
+    public function __construct(
+        iterable $guardrails,
+    ) {
+        $this->guardrails = $guardrails instanceof \Traversable
+            ? iterator_to_array($guardrails)
+            : $guardrails;
+    }
```
Comment (chr-hertel) [inline src/agent/src/Guardrail/GuardrailInputProcessor.php]: shouldn't this work as well?

```suggestion
    /**
     * @param iterable<InputGuardrailInterface> $guardrails
     */
    public function __construct(
        private readonly iterable $guardrails,
    ) {
    }
```

File: src/agent/src/Guardrail/GuardrailOutputProcessor.php [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * @param iterable<OutputGuardrailInterface> $guardrails
+     */
+    public function __construct(
+        iterable $guardrails,
+    ) {
+        $this->guardrails = $guardrails instanceof \Traversable
+            ? iterator_to_array($guardrails)
+            : $guardrails;
+    }
```
Comment (chr-hertel) [inline src/agent/src/Guardrail/GuardrailOutputProcessor.php]:
```suggestion
    /**
     * @param iterable<OutputGuardrailInterface> $guardrails
     */
    public function __construct(
        private readonly iterable $guardrails,
    ) {
    }
```

File: src/agent/src/Guardrail/Scanner/RegexScanner.php (hunk 6 above)
Comment (OskarStark) [reply src/agent/src/Guardrail/Scanner/RegexScanner.php:30]: Or only "guard"?

File: src/agent/src/Guardrail/InputGuardrailInterface.php (hunk 8 above)
Comment (Guikingone) [reply src/agent/src/Guardrail/InputGuardrailInterface.php:26]: Ah yes, forgot about this use case 🙁

File: src/agent/src/Exception/GuardrailException.php (hunk 1 above)
Comment (daifma) [reply src/agent/src/Exception/GuardrailException.php]: DONE

File: src/agent/CHANGELOG.md (hunk 2 above)
Comment (daifma) [reply src/agent/CHANGELOG.md]: Cleaned

File: src/agent/src/Guardrail/GuardrailInputProcessor.php (hunk 3 above)
Comment (daifma) [reply src/agent/src/Guardrail/GuardrailInputProcessor.php:44]: Same pattern as `VoterInterface` voters return a vote, `AccessDecisionManager` decides. Here scanners return a result, the processor decides what to do with it.

File: src/agent/src/Guardrail/GuardrailResult.php (hunk 5 above)
Comment (daifma) [reply src/agent/src/Guardrail/GuardrailResult.php]: Done

File: src/agent/src/Guardrail/Scanner/RegexScanner.php (hunk 6 above)
Comment (daifma) [reply src/agent/src/Guardrail/Scanner/RegexScanner.php:30]: **Guardrail** comes from [OpenAI Agents SDK](https://openai.github.io/openai-agents-python/guardrails/ ) their classes are `InputGuardrail`, `OutputGuardrail`, `InputGuardrailResult`. It's become the standard term in the AI agent space.
Happy to rename to whatever @chr-hertel prefers.

File: src/agent/src/Guardrail/GuardrailInputProcessor.php (hunk 9 above)
Comment (daifma) [reply src/agent/src/Guardrail/GuardrailInputProcessor.php]: Done,

File: src/agent/src/Guardrail/GuardrailOutputProcessor.php (hunk 10 above)
Comment (daifma) [reply src/agent/src/Guardrail/GuardrailOutputProcessor.php]: Done,

File: src/agent/src/Guardrail/Scanner/RegexScanner.php (hunk 7 above)
Comment (daifma) [reply src/agent/src/Guardrail/Scanner/RegexScanner.php:48]: The `@preg_match($pattern, '')` here only validates that the regex is syntactically correct in the constructor.
`symfony/string` is not a direct dependency of the Agent component, @Guikingone would using `UnicodeString` here justify adding it?

Comment (OskarStark) [reply src/agent/src/Guardrail/Scanner/RegexScanner.php:48]: Lets not add this dep yet

File: src/agent/src/Guardrail/Scanner/RegexScanner.php (hunk 6 above)
Comment (OskarStark) [reply src/agent/src/Guardrail/Scanner/RegexScanner.php:30]: Ok then lets go with Guardrail for now.

**Comment by OskarStark** [comment]
Could you please add 2 examples? One for Input and one for output validation and documentation is needed.

AI Bundle integration could be a follow up in my opinion

File: docs/components/agent.rst [hunk 11]
```diff
@@ -564,6 +564,240 @@ in case the processor implemented the :class:`Symfony\\AI\\Agent\\AgentAwareInte
...
+execution halts.
+
+Scanners are executed in registration order. The first scanner that returns a
+:class:`Symfony\\AI\\Agent\\Guardrail\\GuardrailResult` with ``isTriggered() === true`` stops the pipeline
+immediately — remaining scanners are skipped.
+
+Built-in Scanners
+~~~~~~~~~~~~~~~~~
+
+The component ships with three scanners:
```
Comment (OskarStark) [inline docs/components/agent.rst]:
```suggestion
The component ships with the following scanners:
```

**Review by OskarStark — APPROVED** [review]

File: docs/components/agent.rst [hunk 12]
```diff
@@ -564,6 +564,240 @@ in case the processor implemented the :class:`Symfony\\AI\\Agent\\AgentAwareInte
...
+Scanners are executed in registration order. The first scanner that returns a
+:class:`Symfony\\AI\\Agent\\Guardrail\\GuardrailResult` with ``isTriggered() === true`` stops the pipeline
+immediately — remaining scanners are skipped.
+
+Built-in Scanners
+~~~~~~~~~~~~~~~~~
+
+The component ships with the following scanners:
+
+:class:`Symfony\\AI\\Agent\\Guardrail\\Scanner\\PromptInjectionScanner`
```
Comment (OskarStark) [inline docs/components/agent.rst]: instead of using a list here, lets use dedicated sections with subheadlines. advantage is, someone can directly link to this section.

File: docs/components/agent.rst [hunk 13]
```diff
@@ -564,6 +564,240 @@ in case the processor implemented the :class:`Symfony\\AI\\Agent\\AgentAwareInte
...
+        $piiDetector = new RegexScanner(
+            patterns: [
+                '/[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/',                       // Email
+                '/\b(?:\+?\d{1,3}[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b/',         // Phone
+            ],
+            reason: 'Response contains PII (email or phone number).',
+        );
+
+Wiring Guardrails to an Agent
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (OskarStark) [inline docs/components/agent.rst]:
```suggestion
Wiring Guardrails to Agents
~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: docs/components/agent.rst [hunk 14]
```diff
@@ -564,6 +564,240 @@ in case the processor implemented the :class:`Symfony\\AI\\Agent\\AgentAwareInte
...
+            if (!\is_string($content)) {
+                return GuardrailResult::pass();
+            }
+
+            foreach ($this->competitors as $name) {
+                if (str_contains(strtolower($content), strtolower($name))) {
+                    return GuardrailResult::block(
+                        'competitor_mention',
+                        \sprintf('Response mentions competitor "%s".', $name),
+                        1.0,
```
Comment (OskarStark) [inline docs/components/agent.rst]: this is the default, right? lets keep the examples concise

**Comment by OskarStark** [comment]
Should be mentioned here IMHO: https://symfony.com/doc/current/ai/index.html#key-features

File: docs/components/agent.rst (hunk 14 above)
Comment (daifma) [reply docs/components/agent.rst]: Yes, removed.

File: src/agent/src/Guardrail/Scanner/InvisibleCharacterScanner.php [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+ * @author Abderrahman Daif <daif.abderrahman@gmail.com>
+ */
+final class InvisibleCharacterScanner implements InputGuardrailInterface
+{
+    private const SCANNER_NAME = 'invisible_characters';
+
+    /**
+     * Unicode patterns for invisible/zero-width characters.
+     *
+     * @see https://unicode-explorer.com/articles/invisible-characters
```
Comment (chr-hertel) [inline src/agent/src/Guardrail/Scanner/InvisibleCharacterScanner.php:35]: since this URL raises a 404 for me, I wonder how accurate the `const INVISIBLE_PATTERN` is

Comment (DZunke) [reply src/agent/src/Guardrail/Scanner/InvisibleCharacterScanner.php:35]: For me the pattern looks really good, even the URL should maybe be fixed to an existing ressource. 

One of the most importent pattern for hiding messages is covered in there ... 

https://www.promptfoo.dev/blog/invisible-unicode-threats/ 😄

**Comment by chr-hertel** [comment]
Thanks again for working on this @daifma - we had some discussions about the scope of v1.0 and will very likely not merge more sophisticated features like this one in yet. not saying that future versions/packages might not tackle this topic.

if you can't wait for that, and you want to share your work with the community, releasing a standalone package that sits on top of `symfony/ai-agent` would be a great way to got :+1: 

Thanks again & cheers!

**Comment by chr-hertel** [comment]
Closing here for now - cookbook article would be welcome about best practices for preventing prompt injection tho.
